### PR TITLE
Refactor and cleanup gossipsub

### PR DESCRIPTION
**Pull request by NIC619**
## How was it fixed?

Mainly refactoring the Gossipsub `heartbeat`.
- some refactors, remove unnecessary check and peer selection filter
- don't send the control messages at once but instead group the messages for the same peer and send it at the end of the heartbeat
- add tests for individual `mesh_heartbeat` and `gossip_heartbeat`

And some other bug fixes and remove outdated `TODO`s.

#### Cute Animal Picture

(source: https://www.maxpixels.net/Mammals-Cute-Pet-Dog-Animal-Kingdom-Portrait-3071189)

### CODE REVIEW

File: libp2p/pubsub/gossipsub.py [hunk 1]
```diff
@@ -106,7 +109,6 @@ def attach(self, pubsub: Pubsub) -> None:
         logger.debug("attached to pusub")
 
         # Start heartbeat now that we have a pubsub instance
-        # TODO: Start after delay
```
Comment (NIC619) [inline libp2p/pubsub/gossipsub.py]: This is fixed by adding `heartbeat_initial_delay`.

File: libp2p/pubsub/gossipsub.py [hunk 2]
```diff
@@ -127,8 +129,9 @@ def add_peer(self, peer_id: ID, protocol_id: TProtocol) -> None:
             #   instance in multistream-select, but it is not the protocol that gossipsub supports.
             #   In this case, probably we registered gossipsub to a wrong `protocol_id`
             #   in multistream-select, or wrong versions.
-            # TODO: Better handling
```
Comment (NIC619) [inline libp2p/pubsub/gossipsub.py]: Not much to be handled here. This code base should not be reached.

File: libp2p/pubsub/gossipsub.py [hunk 3]
```diff
@@ -218,15 +235,9 @@ def _get_peers_to_send(
 
             # gossipsub peers
             in_topic_gossipsub_peers: List[ID] = None
-            # TODO: Do we need to check `topic in self.pubsub.my_topics`?
```
Comment (NIC619) [inline libp2p/pubsub/gossipsub.py]: Checked in L226

File: libp2p/pubsub/gossipsub.py [hunk 4]
```diff
@@ -344,18 +435,31 @@ def _get_peers_to_send(
...
+        return peers_to_graft, peers_to_prune
 
-    async def fanout_heartbeat(self) -> None:
+    def fanout_heartbeat(self) -> None:
         # Note: the comments here are the exact pseudocode from the spec
         for topic in self.fanout:
-            # If time since last published > ttl
+            # Delete topic entry if it's not in `pubsub.peer_topics`
+            # or if it's time-since-last-published > ttl
             # TODO: there's no way time_since_last_publish gets set anywhere yet
```
Comment (NIC619) [inline libp2p/pubsub/gossipsub.py]: This will be handled in another PR.

File: libp2p/pubsub/gossipsub.py [hunk 5]
```diff
@@ -369,51 +473,44 @@ def _get_peers_to_send(
                     # Add the peers to fanout[topic]
                     self.fanout[topic].extend(selected_peers)
 
-    async def gossip_heartbeat(self) -> None:
+    def gossip_heartbeat(self) -> Dict[ID, Dict[str, List[str]]]:
+        peers_to_gossip: Dict[ID, Dict[str, List[str]]] = {}
         for topic in self.mesh:
             msg_ids = self.mcache.window(topic)
             if msg_ids:
-                # TODO: Make more efficient, possibly using a generator?
```
Comment (NIC619) [inline libp2p/pubsub/gossipsub.py]: Seems irrelevant now IMO.

File: libp2p/pubsub/pubsub.py [hunk 6]
```diff
@@ -78,7 +78,6 @@ class Pubsub:
 
     topic_validators: Dict[str, TopicValidator]
 
-    # TODO: Be sure it is increased atomically everytime.
```
Comment (NIC619) [inline libp2p/pubsub/pubsub.py]: Seems to be the case.

File: libp2p/pubsub/pubsub.py [hunk 7]
```diff
@@ -300,7 +299,6 @@ def get_msg_validators(self, msg: rpc_pb2.Message) -> Tuple[TopicValidator, ...]
             logger.debug("Fail to add new peer %s: stream closed", peer_id)
             del self.peers[peer_id]
             return
-        # TODO: Check EOF of this stream.
```
Comment (NIC619) [inline libp2p/pubsub/pubsub.py]: It's checked right above.

File: libp2p/pubsub/pubsub.py [hunk 8]
```diff
@@ -318,19 +316,20 @@ def _handle_dead_peer(self, peer_id: ID) -> None:
...
 
         self.router.remove_peer(peer_id)
 
         logger.debug("removed dead peer %s", peer_id)
 
     async def handle_peer_queue(self) -> None:
-        """
-        Continuously read from peer queue and each time a new peer is found,
-        open a stream to the peer using a supported pubsub protocol
-        TODO: Handle failure for when the peer does not support any of the
```
Comment (NIC619) [inline libp2p/pubsub/pubsub.py]: Related to the unreachable code path mentioned above.

**Comment by NIC619** [comment]
This should be ready for review. Tasks like `heartbeat` silently crashed have make debugging the tests fairly harder but this problem should be fixed in #365 once we run each part as a  `service` so that crashed child task would terminate the service as well.

File: tests/pubsub/test_gossipsub.py [hunk 9]
```diff
@@ -148,7 +149,7 @@
     await connect(hosts[index_alice], hosts[index_bob])
 
     # Wait 3 seconds for heartbeat to allow mesh to connect
```
Comment (ChihChengLiang) [inline tests/pubsub/test_gossipsub.py]: Dangling magic number here.
```suggestion
     # Wait for heartbeat to allow mesh to connect
```

File: tests/pubsub/test_gossipsub.py [hunk 10]
```diff
@@ -157,15 +158,17 @@
     # alice emit prune message to bob, alice should be removed
     # from bob's mesh peer
     await gossipsubs[index_alice].emit_prune(topic, id_bob)
+    # `emit_prune` does not remove bob from alice's mesh peers
+    assert id_bob in gossipsubs[index_alice].mesh[topic]
 
     # FIXME: This test currently works because the heartbeat interval
     # is increased to 3 seconds, so alice won't get add back into
```
Comment (ChihChengLiang) [inline tests/pubsub/test_gossipsub.py]: The "3 seconds" magic number here too.

File: libp2p/pubsub/gossipsub.py [hunk 11]
```diff
@@ -292,7 +305,76 @@ def _get_peers_to_send(
...
+                del peers_to_gossip[peer]
+
+            control_msg = self.pack_control_msgs(ihave_msgs, None, prune_msgs)
+            await self.emit_control_message(control_msg, peer)
+
+        # Fianlly IHAVE messages
+        for peer in peers_to_gossip:
+            ihave_msgs = []
+            for topic in peers_to_gossip[peer]:
+                ihave_msg = rpc_pb2.ControlIHave()
```
Comment (mhchia) [inline libp2p/pubsub/gossipsub.py]: IMO you can just initialize `rpc_pb2.ControlIHave`.

```python
ihave_msg = rpc_pb2.ControlIHave(
    messageIDs=peers_to_gossip[peer][topic],
    topicID=topic,
)
```

File: libp2p/pubsub/gossipsub.py [hunk 12]
```diff
@@ -324,15 +417,15 @@ def _get_peers_to_send(
...
-                ]
-                for peer in fanout_peers_not_in_mesh:
+                for peer in selected_peers:
                     # Add peer to mesh[topic]
                     self.mesh[topic].append(peer)
 
                     # Emit GRAFT(topic) control message to peer
-                    await self.emit_graft(topic, peer)
+                    if peer not in peers_to_graft:
+                        peers_to_graft[peer] = [topic]
```
Comment (mhchia) [inline libp2p/pubsub/gossipsub.py]: Would it be better to use `defaultdict` for `peers_to_graft`?

File: libp2p/pubsub/gossipsub.py [hunk 13]
```diff
@@ -344,18 +437,34 @@ def _get_peers_to_send(
                     self.mesh[topic].remove(peer)
 
                     # Emit PRUNE(topic) control message to peer
-                    await self.emit_prune(topic, peer)
+                    if peer not in peers_to_prune:
+                        peers_to_prune[peer] = [topic]
+                    else:
+                        peers_to_prune[peer].append(topic)
```
Comment (mhchia) [inline libp2p/pubsub/gossipsub.py]: so does for `peers_to_prune`?

File: libp2p/pubsub/gossipsub.py [hunk 14]
```diff
@@ -369,51 +478,43 @@ def _get_peers_to_send(
...
                     for peer in peers_to_emit_ihave_to:
-                        # TODO: this line is a monster, can hopefully be simplified
-                        if (
-                            topic not in self.mesh or (peer not in self.mesh[topic])
-                        ) and (
-                            topic not in self.fanout or (peer not in self.fanout[topic])
-                        ):
-                            msg_id_strs = [str(msg_id) for msg_id in msg_ids]
-                            await self.emit_ihave(topic, msg_id_strs, peer)
+                        if peer not in peers_to_gossip:
```
Comment (mhchia) [inline libp2p/pubsub/gossipsub.py]: ditto

File: libp2p/pubsub/gossipsub.py [hunk 15]
```diff
@@ -369,51 +478,43 @@ def _get_peers_to_send(
...
+            if msg_ids:
+                # Get all pubsub peers in topic and only add if they are gossipsub peers also
+                if topic in self.pubsub.peer_topics:
+                    # Select D peers from peers.gossipsub[topic]
+                    peers_to_emit_ihave_to = self._get_in_topic_gossipsub_peers_from_minus(
+                        topic, self.degree, self.fanout[topic]
+                    )
+                    msg_id_strs = [str(msg) for msg in msg_ids]
+                    for peer in peers_to_emit_ihave_to:
+                        if peer not in peers_to_gossip:
```
Comment (mhchia) [inline libp2p/pubsub/gossipsub.py]: ditto for defaultdict

File: libp2p/pubsub/gossipsub.py [hunk 16]
```diff
@@ -548,6 +649,21 @@ def _get_in_topic_gossipsub_peers_from_minus(
 
     # RPC emitters
 
+    def pack_control_msgs(
+        self,
+        ihave_msgs: List[rpc_pb2.ControlIHave],
+        graft_msgs: List[rpc_pb2.ControlGraft],
+        prune_msgs: List[rpc_pb2.ControlPrune],
+    ) -> rpc_pb2.ControlMessage:
```
Comment (mhchia) [inline libp2p/pubsub/gossipsub.py]: Am I understanding it correctly that this change is for piggybacking?

File: libp2p/pubsub/gossipsub.py (hunk 12 above)
Comment (NIC619) [reply libp2p/pubsub/gossipsub.py]: Good idea

File: libp2p/pubsub/gossipsub.py (hunk 16 above)
Comment (NIC619) [reply libp2p/pubsub/gossipsub.py]: I would say it's just packing the messages together and send them at once. Will open an issue for piggybacking.

File: libp2p/pubsub/gossipsub.py [hunk 17]
```diff
@@ -204,42 +215,46 @@ def _get_peers_to_send(
...
+                # as our `fanout` peers.
+                topic_in_fanout: bool = topic in self.fanout
+                fanout_peers: List[ID] = self.fanout[topic] if topic_in_fanout else []
+                fanout_size = len(fanout_peers)
+                if not topic_in_fanout or (
+                    topic_in_fanout and fanout_size < self.degree
+                ):
+                    if topic in self.pubsub.peer_topics:
+                        # Combine fanout peers with selected peers
+                        fanout_peers += self._get_in_topic_gossipsub_peers_from_minus(
```
Comment (mhchia) [inline libp2p/pubsub/gossipsub.py]: Note: one thing I'm thinking is to have methods like `_get_fanout_peers(topic)`, `_get_mesh_peers(topic)`, and `_get_ peers_to_emit_ihave_to`, to make the code more readable.

**Comment by moshemalawach** [comment]
Very nice! I used gossipsub in my implementation before... TL;DR, what should I change?

**Comment by ralexstokes** [comment]
@moshemalawach i'm not sure a user of the code in this PR will need to update anything beyond any API changes introduced.

note that we will be migrating to `gossipsub v1.1` over the next few months -- v1.1 is backwards compatible w/ v1.0 but does add a variety of "hardening" features to make the overall mesh more secure.
